vendor/ocaml/bytecomp/matching.ml

```diff
@@ -2815,6 +2815,19 @@ and do_compile_matching repr partial ctx arg pmh = match pmh with
         divide_constant
         (combine_constant pat.pat_loc arg cst partial)
         ctx pm
+  | Tpat_construct ({txt = Longident.Lident "false"}, cstr, _)
+  | Tpat_construct ({txt = Longident.Lident "true"}, cstr, _) ->
+      let lambda, cont =
+        compile_test
+          (compile_match repr partial) partial
+          divide_constructor (combine_constructor pat.pat_loc arg pat cstr partial)
+          ctx pm in
+      let lambda1 = match lambda with
+        | Lifthenelse (Lprim (Pintcomp Cneq, [l1; Lconst (Const_base (Const_int n))], loc), l2, l3) ->
```

Comment:
Counter-example showing when a single case is considered in the compilation of pattern matching, then the "when" clause is emitted directly without any further case analysis, and here it's wrongly interpreted as a boolean case analysis:
```
let f v k = match v with
| (_, 1, true) when k>0 -> 111
| (_, 2, true) when k < 8 -> 222
| (false, 2, true)
  (* XXX fools replacement of (k != 0) into thinking this is a bool check *)
  when k != 0 -> 333
| (false, 3, true)
  (* XXX not matched *)
  when k > 77 -> 444
| (_, _, false) -> 555
```